Remove hardcoded shortcuts from /scene and instead use the input action system to allow them to be customised.

EricEzaM opened the pull request:
* Made shortcuts which were previously hardcoded customisable. (this is just the backend implementation - the UI to do so will come in a later PR)
* Shortcuts that fall into this category are things like cut, copy paste, etc on TextEdit. In the code, these were defined in `gui_input` with things like 
`if (k->is_pressed() && k->get_keycode() == KEY_C && k->get_command() == true){ /* copy */}`
* Cleaned up logic... it is much more readable now. The below code example handles 1) moving the cursor left 1 char, 2) moving the cursor left by 1 word 3) moving the cursor to the start of the line (on Apple only) and finally 4) doing the previous 3 with selection (shift) pressed.

https://github.com/godotengine/godot/blob/07112b92a85de9b2134f3208149b295f871d3642/scene/gui/text_edit.cpp#L2876-L2945

Quite hard to follow in my opinion. This logic now looks like this, except the following logic can use any key combination, not just the left arrow key.
```cpp
// In gui_input:
if (k->is_action("ui_text_caret_left")) {
	_move_cursor_left(shift_pressed, false);
}
if (k->is_action("ui_text_caret_word_left")) {
	_move_cursor_left(shift_pressed, true);
}
// …
if (k->is_action("ui_text_caret_line_start")) {
	_move_cursor_to_line_start(shift_pressed);
}
// Other methods in the class:
void TextEdit::_move_cursor_left(bool p_select, bool p_move_by_word);
void TextEdit::_move_cursor_to_line_start(bool p_select);
```

The keys used to perform these actions are defined in InputMap:
```cpp
inputs = List<Ref<InputEvent>>();
inputs.push_back(InputEventKey::create_reference(KEY_HOME));
map.insert("ui_text_caret_line_start", inputs);

inputs = List<Ref<InputEvent>>();
inputs.push_back(InputEventKey::create_reference(KEY_A, KEY_MASK_CTRL));
inputs.push_back(InputEventKey::create_reference(KEY_LEFT, KEY_MASK_CMD));
map.insert("ui_text_caret_line_start.OSX", inputs);
```

**Added Utility Method**
Added static methods `::create_reference` on InputEventKey and JoypadButton which is a one-line way to do:
```
Ref<InputEventJoypadButton> ie;
ie.instance();
ie->set_button_index(p_btn_index);
```
The above code is used in many places with Buttons and Keys, so moving it to a utility one-liner saves A LOT of lines of code, at least a couple hundred lines.

**Unified definitions of the same shortcuts**
Moved builtin input actions away from being defined twice in ProjectSettings and EditorSettings to only being defined once in InputMap. Project/Editor Settings instead just get the builtin actions and loop through them to add them to their settings list. Reduces code duplication.

Closes #17927
Closes #29490
Closes #12164
Closes #42139
Closes #26854
Closes godotengine/godot-proposals#1532

This is the fourth PR of a larger series based on #42600.

EricEzaM commented:
Requesting review from @Paulb23 as we had quite the dialog about this PR in the associated proposal thread.

akien-mga commented:
For the reference, here's the list of default actions in the InputMap as of this PR. I know #42600 might refine this further but this might still be worth a discussion on whether all of those should really be configurable (like caret up/right bound to <kbd>Up</kbd>/<kbd>Right</kbd>), or if they should be renamed/better sorted to make it more obvious what is being affected by those.

My main concern seeing this list is that it's huge for something that most users won't even want to configure. On the other hand the same can likely be said of some of the original `ui_` actions prior to this. (Though many users tend to reuse `ui_<direction>` and `ui_accept` for their game logic on simple projects.)

So IMO it's most important to ensure that the InputMap editor in the Project Settings will be as user-friendly as possible, even more so than the Editor Settings' shortcut configuration tab which is more advanced / less critical to each project. Again, this might be handled in #42600 already but discussion there seemed to focus on the Editor Settings. I'll give it a try to see.

User-defined actions and advanced UI configuration should probably be separated so that users can focus on their game actions first and foremost, and cater to fine tuning of Control actions only when they needed it (very rare cases IMO).

![Screenshot_20201123_115805](https://user-images.githubusercontent.com/4701338/99954491-27679e80-2d83-11eb-82dc-a5cddefedc01.png)
![Screenshot_20201123_115834](https://user-images.githubusercontent.com/4701338/99954532-38b0ab00-2d83-11eb-89aa-2327671f8077.png)
![Screenshot_20201123_115849](https://user-images.githubusercontent.com/4701338/99954561-423a1300-2d83-11eb-8626-7a31f9f83c10.png)
![Screenshot_20201123_115905](https://user-images.githubusercontent.com/4701338/99954585-4a924e00-2d83-11eb-92de-4c274b99e2c3.png)
![Screenshot_20201123_115936](https://user-images.githubusercontent.com/4701338/99954625-5da51e00-2d83-11eb-9e19-ff58f3b4bce2.png)

EricEzaM commented:
@akien-mga There should also be a full list at input_map.cpp @ line 267

Here I put the builtin action display names, for cases where they need to be displayed in a popup menu (like cut, copy paste, etc).
```
static const _BuiltinActionDisplayName _builtin_action_display_names[] = {
	/* clang-format off */
	{ "ui_accept", 									"Accept" },
	{ "ui_select", 									"Select" },
	{ "ui_cancel", 									"Cancel" },
	{ "ui_focus_next", 								"Focus Next" },
	{ "ui_focus_prev", 								"Focus Prev" },
	{ "ui_left", 									"Left" },
	{ "ui_right", 									"Right" },
	{ "ui_up", 										"Up" },
	{ "ui_down", 									"Down" },
	{ "ui_page_up",									"Page Up" },
	{ "ui_page_down", 								"Page Down" },
	{ "ui_home", 									"Home" },
	{ "ui_end", 									"End" },
	{ "ui_cut", 									"Cut" },
	{ "ui_copy", 									"Copy" },
	{ "ui_paste", 									"Paste" },
	{ "ui_undo", 									"Undo" },
	{ "ui_redo", 									"Redo" },
	{ "ui_text_completion_query", 					"Completion Query" },
	{ "ui_text_newline", 							"New Line" },
	{ "ui_text_newline_blank",						"New Blank Line" },
	{ "ui_text_newline_above", 						"New Line Above" },
	{ "ui_text_indent", 							"Indent" },
	{ "ui_text_dedent", 							"Dedent" },
	{ "ui_text_backspace", 							"Backspace" },
	{ "ui_text_backspace_word", 					"Backspace Word" },
	{ "ui_text_backspace_word.OSX", 				"Backspace Word" },
	{ "ui_text_backspace_all_to_left", 				"Backspace all to Left" },
	{ "ui_text_backspace_all_to_left.OSX", 			"Backspace all to Left" },
	{ "ui_text_delete", 							"Delete" },
	{ "ui_text_delete_word", 						"Delete Word" },
	{ "ui_text_delete_word.OSX", 					"Delete Word" },
	{ "ui_text_delete_all_to_right", 				"Delete all to Right" },
	{ "ui_text_delete_all_to_right.OSX", 			"Delete all to Right" },
	{ "ui_text_caret_left", 						"Caret Left" },
	{ "ui_text_caret_word_left", 					"Caret Word Left" },
	{ "ui_text_caret_word_left.OSX", 				"Caret Word Left" },
	{ "ui_text_caret_right", 						"Caret Right" },
	{ "ui_text_caret_word_right", 					"Caret Word Right" },
	{ "ui_text_caret_word_right.OSX", 				"Caret Word Right" },
	{ "ui_text_caret_up", 							"Caret Up" },
	{ "ui_text_caret_down", 						"Caret Down" },
	{ "ui_text_caret_line_start", 					"Caret Line Start" },
	{ "ui_text_caret_line_start.OSX", 				"Caret Line Start" },
	{ "ui_text_caret_line_end", 					"Caret Line End" },
	{ "ui_text_caret_line_end.OSX",					"Caret Line End" },
	{ "ui_text_caret_page_up", 						"Caret Page Up" },
	{ "ui_text_caret_page_down", 					"Caret Page Down" },
	{ "ui_text_caret_document_start", 				"Caret Document Start" },
	{ "ui_text_caret_document_start.OSX",			"Caret Document Start" },
	{ "ui_text_caret_document_end", 				"Caret Document End" },
	{ "ui_text_caret_document_end.OSX", 			"Caret Document End" },
	{ "ui_text_scroll_up", 							"Scroll Up" },
	{ "ui_text_scroll_up.OSX", 						"Scroll Up" },
	{ "ui_text_scroll_down", 						"Scroll Down" },
	{ "ui_text_scroll_down.OSX", 					"Scroll Down" },
	{ "ui_text_select_all", 						"Select All" },
	{ "ui_text_toggle_insert_mode", 				"Toggle Insert Mode" },
	{ "ui_graph_duplicate", 						"Duplicate Nodes" },
	{ "ui_graph_delete", 							"Delete Nodes" },
	{ "ui_filedialog_up_one_level", 				"Go Up One Level" },
	{ "ui_filedialog_refresh", 						"Refresh" },
	{ "ui_filedialog_show_hidden", 					"Show Hidden" },
	{ "",											""}
	/* clang-format on */
};
```

EricEzaM commented:
Hmm those descriptions will also need a `TTRC()` I think

Paulb23 reviewed (commented):
Code looks fine to me, I think this level of configuration is okay, mainly comes down to having a suitable UX, which is easier said then done. The other point is needing someway to document or as perhaps as part of the UX, that appending `.OSX` overrides it for apple keyboards. Though similar to Aiken suggestion in #42600, it might be worth doing the UX changes as a separate proposal / PR so we don't get bogged down here.

Needs a hefty rebase after #41100, which also adds a new shortcut `ui_swap_input_direction = KEY_QUOTELEFT` to `LineEdit` and `TextEdit`.

Lastly, found #22927 which this should close.

EricEzaM commented:
Rebased.

I added the `ui_swap_input_direction` action mapping and created a method for it in LineEdit and TextEdit.

EricEzaM committed:
Added convenience create_reference methods for Key and JoyButton inputs

EricEzaM committed:
Added ability to override built-in actions for the editor

This adds the ability to add overrides for built-in actions (i.e. ui_*) in the editor. Also added a number of additional built-in actions for various text-related actions, gui-generic actions (like copy and paste) and graph-related actions (duplicate nodes), etc. Moved the definition of input actions to input_map, rather than in project_settings so the editor can make use of these actions as well.

EricEzaM committed:
Removed hardcoded shortcuts from /scene and converted to input actions

This removes hardcoded actions from things like LineEdit and TextEdit.

Previously, things like copy, paste, etc were all hardcoded to Ctrl+C, Ctrl+V, etc. They could not be changed. This allows the possibility of them being changed, by making them use the action map. This has the added benefit of greatly simplifying the input handling logic in those controls. The logic which was previously in a huge and hard to follow switch statement has been extracted to individual methods.

Paulb23 reviewed (dismissed):
Got round to testing this locally, seem like we need to fix or workaround #18793 first otherwise we can't determine the difference between `Tab` and `Ctrl+Tab` etc. 

Also spotted a couple more editor shortcuts need updating [here.](https://github.com/godotengine/godot/blob/a132e89b6aca09bd0f0871dca91c1aa648dbad6a/editor/plugins/text_editor.cpp#L587-L608)

EricEzaM commented:
> Got round to testing this locally, seem like we need to fix or workaround #18793 first otherwise we can't determine the difference between `Tab` and `Ctrl+Tab` etc.

This is caused by #44180

Paulb23 commented:
While testing this with #44355, it looks like some changes got lost with the #41100 rebase. `_move_cursor_left` moves the caret right by word and `_move_cursor_right` has not been updated to use `TextServer`.

EricEzaM commented:
Fixed above issues @Paulb23.

I tested every single shortcut in TextEdit and LineEdit, and they are working correctly. Yes, #18793 broke it, however if you order the shortcuts correctly, it will work as intended. For example, take the `ui_text_newline_*` actions...

```
ui_text_newline_above = Control + Shift + Enter
ui_text_newline_blank = Control + Enter
ui_text_newline = Enter
```

If they are placed in order of specificity (as above), then it works ok. Same with Tab and Control+Tab. However, since these shortcuts are customisable, this is **not** an acceptable fix - it is only acceptable for testing the default options. #44355 or a similar fix to this issue will be required.

Note I also added a new `ui_` action, `ui_text_completion_accept`, because previously it was only using `ui_accept`, but we want to use `KEY_TAB` to accept a completion too, but `KEY_TAB` shouldn't be in `ui_accept`.

akien-mga commented:
Will need a rebase after recent changes to TextEdit.

EricEzaM commented:
Rebased

EricEzaM commented:
I have added the use of `exact_match` and tested all the actions in the script editor, and it is working well!

akien-mga commented on core/input/input_map.cpp:
```diff
@@ -260,84 +260,444 @@ void InputMap::load_from_project_settings() {
+
+static const _BuiltinActionDisplayName _builtin_action_display_names[] = {
+	/* clang-format off */
+	{ "ui_accept", 									TTRC("Accept") },
+	{ "ui_select", 									TTRC("Select") },
+	{ "ui_cancel", 									TTRC("Cancel") },
+	{ "ui_focus_next", 								TTRC("Focus Next") },
+	{ "ui_focus_prev", 								TTRC("Focus Prev") },
+	{ "ui_left", 									TTRC("Left") },
+	{ "ui_right", 									TTRC("Right") },
+	{ "ui_up", 										TTRC("Up") },
+	{ "ui_down", 									TTRC("Down") },
+	{ "ui_page_up",									TTRC("Page Up") },
+	{ "ui_page_down", 								TTRC("Page Down") },
+	{ "ui_home", 									TTRC("Home") },
+	{ "ui_end", 									TTRC("End") },
+	{ "ui_cut", 									TTRC("Cut") },
+	{ "ui_copy", 									TTRC("Copy") },
+	{ "ui_paste", 									TTRC("Paste") },
+	{ "ui_undo", 									TTRC("Undo") },
+	{ "ui_redo", 									TTRC("Redo") },
+	{ "ui_text_completion_query", 					TTRC("Completion Query") },
+	{ "ui_text_newline", 							TTRC("New Line") },
+	{ "ui_text_newline_blank",						TTRC("New Blank Line") },
+	{ "ui_text_newline_above", 						TTRC("New Line Above") },
+	{ "ui_text_indent", 							TTRC("Indent") },
+	{ "ui_text_dedent", 							TTRC("Dedent") },
+	{ "ui_text_backspace", 							TTRC("Backspace") },
+	{ "ui_text_backspace_word", 					TTRC("Backspace Word") },
+	{ "ui_text_backspace_word.OSX", 				TTRC("Backspace Word") },
+	{ "ui_text_backspace_all_to_left", 				TTRC("Backspace all to Left") },
+	{ "ui_text_backspace_all_to_left.OSX", 			TTRC("Backspace all to Left") },
+	{ "ui_text_delete", 							TTRC("Delete") },
+	{ "ui_text_delete_word", 						TTRC("Delete Word") },
+	{ "ui_text_delete_word.OSX", 					TTRC("Delete Word") },
+	{ "ui_text_delete_all_to_right", 				TTRC("Delete all to Right") },
```
Aligning code with tabs is never a good option :)
![Screenshot_20210216_100307](https://user-images.githubusercontent.com/4701338/108040903-2c648300-703e-11eb-9692-5a69c5449bb8.png)

Either use spaces to ensure exact alignment, or remove the alignment.

EricEzaM commented:
All done, changed to spaces :)

akien-mga commented:
Needs another rebase after #46127 (then can be merged once CI passes).

akien-mga commented:
Thanks!

EricEzaM commented:
It's in! How exciting!

I'm really happy with how this simplifies the logic in gui input for text and line edit

christinoleo commented:
@EricEzaM Thanks for the work here.
I have a question to you. It seems that due to commits 074f535 and 49714b0 (maybe there are more, but this is as far as I got), by default the autocompletion in the editor accepts Space, Enter and KP_Enter. The space one is causing (at least me) to always when editing a gdscript to autocomplete AABB when typing "<variable> = ". The last space autocompletes to "<variable> =AABB" instead of leaving a space. Of course I can just remove space from the ui_accept shortcuts, but it took me some days to figure this out since godot 3.2 does not have this issue (at least by default). In your opinion, do you think it's worth not having the spacebar as a default ui_accept or is there some other code change which is actually causing this weird autocomplete do be done? I intended to create a new issue ticket but I wish to get your opinion first.. 
![aaaa](https://user-images.githubusercontent.com/1745901/113465366-cecba080-9409-11eb-8835-7694a663c7f2.gif)
Thanks!

EricEzaM commented:
@christinoleo Thanks for pointing this out! Perhaps a new action needs to be created, specifically for accepting auto-complete options, which only includes tab, enter and kp_enter.

Calinou commented:
> @christinoleo Thanks for pointing this out! Perhaps a new action needs to be created, specifically for accepting auto-complete options, which only includes tab, enter and kp_enter.

How should we implement this for use in the editor? Having to edit project-specific settings to tweak autocompletion to your liking sounds bad. Personally, I'd prefer to have only <kbd>Tab</kbd> accept suggestions, but I understand most people want <kbd>Enter</kbd> to accept suggestions as well.

EricEzaM commented:
@Calinou never mind actually - it is a very easy fix. For some reason I check for both `ui_text_completion_accept` and `ui_accept` in the code... It should really only be the former.

https://github.com/godotengine/godot/blob/ed2f51b15fee22a35d2a975fd77a70fc291bd8de/scene/gui/text_edit.cpp#L3262-L3266

YuriSizov commented:
Which shortcut have you modified, and what exactly are you trying to do to test it?

Thane5 commented:
I set "ui_cut" to ctrl+shift+x
![image](https://user-images.githubusercontent.com/13683581/220588158-a261480b-9c9e-497c-ba4b-5b061a481930.png)

For testing, i created a new project. Created a "control" scene, added a button. 

If i select it and press ctrl+shift+x, nothing happens. If i press ctrl+x, the object gets cut.

akien-mga commented:
Works fine for me on Linux with 4.0 RC 3:

![image](https://user-images.githubusercontent.com/4701338/220589750-9ec650c5-99b3-49a4-9b2e-9f1ce3a0dc22.png)

Selecting text in the script editor and pressing Ctrl+Shift+X cuts it. Pressing Ctrl+X writes "x".

Thane5 commented:
In the text editor, it works for me too, but not in the 2D or 3D editor.

Thane5 commented:
Oh, i just realised that there is a seperate "cut" command for the scene tree. I did not find it at first since i use the german version of the editor and unlike "ui_cut", the "scene tree cut" has already been translated into german.

However i find it a bit weird that "ui_undo" and "ui_redo" can affect the scene tree as expected, while "ui_cut" is specific to the text editor.

Anyways, at least my issue is solved now.
